kornia/feature/sold2/sold2.py

```diff
@@ -0,0 +1,368 @@
+        'junction_refine_cfg': {
+            'num_perturbs': 9,
+            'perturb_interval': 0.25,
+        },
+    },
+    'line_matcher_cfg': {
+        'cross_check': True,
+        'num_samples': 5,
+        'min_dist_pts': 8,
+        'top_k_candidates': 10,
+        'grid_size': 4,
+    },
+}
+
+
+class SOLD2(nn.Module):
+    r"""Module, which detects and describe line segments in an image.
+
+    This is based on the original code from the paper "SOLD²: Self-supervised
+    Occlusion-aware Line Detector and Descriptor". See :cite:`SOLD22021` for more details.
+
+    Args:
+        config: Dict specifying parameters. None will load the default parameters,
+            which are tuned for images in the range 400~800 px.
+        pretrained: If True, download and set pretrained weights to the model.
+
+    Returns:
+        The raw junction and line heatmaps, the semi-dense descriptor map,
+        as well as the list of detected line segments (ij coordinates convention).
+
+    Example:
+        >>> images = torch.rand(2, 1, 512, 512)
+        >>> sold2 = SOLD2()
+        >>> outputs = sold2(images)
+        >>> line_seg1 = outputs["line_segments"][0].detach().cpu().numpy()
+        >>> line_seg2 = outputs["line_segments"][1].detach().cpu().numpy()
+        >>> desc1 = outputs["dense_desc"][0]
+        >>> desc2 = outputs["dense_desc"][1]
+        >>> matches = sold2.match(line_seg1, line_seg2, desc1[None], desc2[None])
+    """
+
+    def __init__(self, pretrained: bool = True, config: Dict = None):
+        super().__init__()
+        # Initialize some parameters
+        self.config = default_cfg if config is None else config
+        self.grid_size = self.config["grid_size"]
+        self.junc_detect_thresh = self.config.get("detection_thresh", 1 / 65)
+        self.max_num_junctions = self.config.get("max_num_junctions", 500)
+
+        # Load the pre-trained model
+        self.model = SOLD2Net(self.config)
+        if pretrained:
+            pretrained_dict = torch.hub.load_state_dict_from_url(
+                urls["wireframe"], map_location=lambda storage, loc: storage)
+            state_dict = self.adapt_state_dict(pretrained_dict['model_state_dict'])
+            self.model.load_state_dict(state_dict)
+        self.eval()
+
+        # Initialize the line detector
+        self.line_detector_cfg = self.config["line_detector_cfg"]
+        self.line_detector = LineSegmentDetectionModule(**self.config["line_detector_cfg"])
+
+        # Initialize the line matcher
+        self.line_matcher = WunschLineMatcher(**self.config["line_matcher_cfg"])
+
+    def forward(self, img: torch.Tensor) -> Dict:
+        """
+        Args:
+            img: batched images with shape :math:`(N, 1, H, W)`.
+
+        :return:
+            - ``line_segments``: list of line segments in each of the N images :math:`List[(Nlines, 2, 2)]`.
+            - ``raw_junc_heatmap``: raw junction heatmap of shape :math:`(N, H, W)`.
```

Comment:
the convention we normally use is`B` for number of batches and in this case `N` could be number of lines ?